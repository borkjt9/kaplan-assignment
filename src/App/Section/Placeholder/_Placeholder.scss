$option-height: 60px;
$option-border-radius: 10px;
$option-background: #F9F9F9;
$placeholder-primary-color: lightgray;
$border-color: lightgray;
$shadow-color: #595959;

.placeholder-container {
  position: relative;
  z-index: 2;
  width: 100%;

  &__number {
    position: absolute;
    color: $placeholder-primary-color;
  }
  &__placeholder {
    height: $option-height;
    width: inherit;
    border-radius: $option-border-radius;
    border: 1px dashed $border-color;
  }
}
