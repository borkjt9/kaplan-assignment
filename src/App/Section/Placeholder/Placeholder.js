/**
 * @file Placeholder
 * @author John Borkowski
 * @version 0.1
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { DropTarget } from 'react-dnd';
import { DRAGGABLE_TYPE_ANSWER } from '../../../redux/constants';
import './_Placeholder.scss';

/**
 * Called in drop below after an answer is 'dropped' in the placeholder.
 *
 * @param {function} positionAnswer - Function fed from App component. Connects to redux store.
 * @param {Object} positionData - Object used to position element. Contains the below keys:
 * @key {Number} fromIndex - Old location of answer.
 * @key {Number} toIndex - New location of answer.
 * @key {Number} activeQuestion - Index of active question.
 * @key {Number} fromSide - Old side. Either 'left' or 'right'.
 * @key {Number} toSide - New side. Either 'left' or right'.
 */
function callPositionAnswer(positionAnswer, positionData) {
  positionAnswer(positionData);
}

/**
 * Used as a parameter to DropTarget. Which is a higher order component from react-dnd library.
 * http://react-dnd.github.io/react-dnd/docs-overview.html
 */
const placeholderTarget = {
  /**
   * Returns true because all placeholders are targets
   * @param {Object} props - The components props, plus added props from react-dnd.
   */
  canDrop() {
    return true;
  },

  /**
   * Is called when a component is 'dropped'.
   */
  drop(props) {
    const {
      fromIndex,
      toIndex,
      activeQuestion,
      fromSide,
      toSide,
      positionAnswer,
    } = props;

    const positionData = {
      toIndex,
      fromIndex,
      activeQuestion,
      toSide,
      fromSide,
    };
    callPositionAnswer(positionAnswer, positionData);
  },
};

/**
 * From react-dnd documentation:
 * For each component that needs to track the drag and drop state,
 * You can define a collecting function that retrieves the relevant bits of it from the monitors.
 * React DnD then takes care of timely calling your collecting function
 * and merging its return value into your components' props.
 */
function collect(connect, monitor) {
  const info = {
    connectDropTarget: connect.dropTarget(),
    isOver: monitor.isOver(),
    canDrop: monitor.canDrop(),
  };
  return info;
}

/**
 * Renders overlay on top of placeholder if dragged answer is hovering over it.
 *
 * @param {String} color Color of the overlay.
 */
function renderOverlay(color) {
  return (
    <div
      style={{
        position: 'absolute',
        top: 0,
        left: 0,
        borderRadius: 10,
        height: '100%',
        width: '104%',
        marginLeft: '-2%',
        zIndex: 2,
        backgroundColor: color,
      }}
    />
  );
}

export class Placeholder extends Component {
  /**
   * Placeholders are rendered in locations whenever an answer is not.
   * Connected to react-dnd higher order component 'DropTarget'.
   */
  render() {
    const {
      connectDropTarget,
      isOver,
      canDrop,
      toIndex,
    } = this.props;

    //connectDropTarget is react-dnd function.
    return connectDropTarget(
      <div className="placeholder-container my-4 d-flex align-items-center">
        <h1 className="placeholder-container__number ml-4">{ toIndex + 1 }</h1>
        <div className="placeholder-container__placeholder" />
        {isOver && canDrop && renderOverlay('#007bff')}
      </div>,
    );
  }
}

Placeholder.propTypes = {
  /** Supplied by react-dnd. */
  connectDropTarget: PropTypes.func.isRequired,
  /** Supplied by react-dnd. True if drag item is hovering over placeholder */
  isOver: PropTypes.bool.isRequired,
  /** Supplied by react-dnd. True for all Placeholders */
  canDrop: PropTypes.bool.isRequired,
  /** Index of placeholder. Used if answer is 'dropped'. */
  toIndex: PropTypes.number.isRequired,
  /** Old answer index. Used if answer is 'dropped'. */
  fromIndex: PropTypes.number.isRequired,
  /** Changes position of answer to current placeholder position. Fed up to App component. */
  positionAnswer: PropTypes.func.isRequired,
  /** Active question number. Used if answer is 'dropped'. */
  activeQuestion: PropTypes.number.isRequired,
  /** side of old answer. Used if answer is 'dropped'. */
  fromSide: PropTypes.string.isRequired,
  /** side of placeholder. Used if answer is 'dropped'. */
  toSide: PropTypes.string.isRequired,
};

export default DropTarget(DRAGGABLE_TYPE_ANSWER, placeholderTarget, collect)(Placeholder);
