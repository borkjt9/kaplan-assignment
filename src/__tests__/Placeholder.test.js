import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';

import { Placeholder } from '../App/Section/Placeholder/Placeholder';

Enzyme.configure({ adapter: new Adapter() });


const identity = el => el;


describe('Placeholder', () => {
  let wrapper;

  it('renders snapshot correctly', () => {
    const renderedValue =  renderer.create(
      <Placeholder
        fromIndex={0}
        toIndex={0}
        fromSide="left"
        toSide="right"
        activeQuestion={0}
        positionAnswer={identity}
        connectDropTarget={identity}
        isOver
        canDrop
      />).toJSON()
    expect(renderedValue).toMatchSnapshot();
  });

  it('renders all placeholder text correctly', () => {
    const indexArr = [0, 1, 2, 3, 4];
    indexArr.forEach((index) => {
      wrapper = shallow(
        <Placeholder
          fromIndex={0}
          toIndex={index}
          fromSide="left"
          toSide="right"
          activeQuestion={0}
          positionAnswer={identity}
          connectDropTarget={identity}
          isOver
          canDrop
        />
      );
      const renderedText = wrapper.text();
      const expectedText = (index + 1).toString();
      expect(renderedText).toEqual(expectedText);
    });
  });
});
