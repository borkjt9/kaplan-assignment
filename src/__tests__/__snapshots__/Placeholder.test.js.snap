// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Placeholder renders snapshot correctly 1`] = `
<div
  className="placeholder-container my-4 d-flex align-items-center"
>
  <h1
    className="placeholder-container__number ml-4"
  >
    1
  </h1>
  <div
    className="placeholder-container__placeholder"
  />
  <div
    style={
      Object {
        "backgroundColor": "#007bff",
        "borderRadius": 10,
        "height": "100%",
        "left": 0,
        "marginLeft": "-2%",
        "position": "absolute",
        "top": 0,
        "width": "104%",
        "zIndex": 2,
      }
    }
  />
</div>
`;
